File: GFG_Delete_Without_Head_Pointer-1.java
// Delete without head pointer - https://practice.geeksforgeeks.org/problems/delete-without-head-pointer/1
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : On geeksforgeeks
// Any problem you faced while coding this : No

/*Time Complexity: 
For printing linked list: O(N)
For inserting node: O(1)
For deleting node: O(N)
Auxiliary Space: O(1) */
/*
class Node
{
	int data ;
	Node next;
	Node(int d)
	{
		data = d;
		next = null;
	}
}
*/

//Function to delete a node without any reference to head pointer.
class Solution
{
    void deleteNode(Node del)
    {
        Node temp = del.next;
        del.data = temp.data;
        del.next = temp.next;
        temp = null;
    }
}

// ----------------------------------------------------------|||||||||||||||||||||||||||------------------------------------------------------- \\

// WHOLE CODE 

class LinkedList {
    Node head; // head of the list

    class Node {
        int data;
        Node next;
        Node(int d)
        {
            data = d;
            next = null;
        }
    }

    /* Given a reference to the head of a list and an int,
        inserts a new Node on the front of the list. */
    public void push(int new_data)
    {
        /* 1. alloc the Node and put the data */
        Node new_Node = new Node(new_data);

        /* 2. Make next of new Node as head */
        new_Node.next = head;

        /* 3. Move the head to point to new Node */
        head = new_Node;
    }

    /* This function prints contents of linked list 
        starting from the given Node */
    public void printList()
    {
        Node tNode = head;
        while (tNode != null) {
            System.out.print(tNode.data + " ");
            tNode = tNode.next;
        }
    }

    public void deleteNode(Node Node_ptr)
    {
        Node temp = Node_ptr.next;
        Node_ptr.data = temp.data;
        Node_ptr.next = temp.next;
        temp = null;
    }

    public static void main(String[] args)
    {
        LinkedList llist = new LinkedList();

        /* Use push() to construct below list
        1->12->1->4->1  */
        llist.push(1);
        llist.push(4);
        llist.push(1);
        llist.push(12);
        llist.push(1);

        System.out.println("Before deleting");
        llist.printList();

        /* I m deleting the head itself.
        You can check for more cases */
        llist.deleteNode(llist.head);

        System.out.println("\nAfter Deleting");
        llist.printList();
    }
}

File: gfg-delete-without-head-pointer-1.py
# https://www.geeksforgeeks.org/problems/delete-without-head-pointer/1


#User function Template for python3
'''
    Your task is to delete the given node from
	the linked list, without using head pointer.
	
	Function Arguments: node (given node to be deleted) 
	Return Type: None, just delete the given node from the linked list.

	{
		# Node Class
		class Node:
		    def __init__(self, data):   # data -> value stored in node
		        self.data = data
		        self.next = None
	}
 
 PROBLEM = #Function to delete a node without any reference to head pointer.
'''
class Solution:
    """
    Accepted
    Time Complexity : O(1)
    Space Complexity : O(1)
    
    Explanation:
        1. Copy the data from next node to current node
        2. Delete the next node like a regular deletion
    """
    def deleteNode(self,curr_node):
        #code here
        tmp = curr_node.next.next
        val = curr_node.next.data
        curr_node.data = val
        curr_node.next.next = None
        curr_node.next = tmp